### PR TITLE
Materialize compact subscripts when slicing ALTREP objects

### LINKED ISSUE #744: vec_rbind() crashes with tibble from vroom()
The following code produces a hard crash on my system with the dev version of `vctrs` (respl `dplyr` and its `bind_rows()` function)

``` r
x <- vroom::vroom("x,y\n1,2\n")
#> Rows: 1
#> Columns: 2
#> Delimiter: ","
#> dbl [2]: x, y
#> 
#> Use `spec()` to retrieve the guessed column specification
#> Pass a specification to the `col_types` argument to quiet this message
vctrs::vec_rbind(x)
```

<sup>Created on 2020-01-14 by the [reprex package](https://reprex.tidyverse.org) (v0.3.0)</sup>

<details>

<summary>Session info</summary>

``` r
devtools::session_info()
#> ─ Session info ───────────────────────────────────────────────────────────────
#>  setting  value                       
#>  version  R version 3.6.1 (2019-07-05)
#>  os       macOS Catalina 10.15.2      
#>  system   x86_64, darwin15.6.0        
#>  ui       X11                         
#>  language (EN)                        
#>  collate  en_US.UTF-8                 
#>  ctype    en_US.UTF-8                 
#>  tz       UTC                         
#>  date     2020-01-14                  
#> 
#> ─ Packages ───────────────────────────────────────────────────────────────────
#>  package     * version      date       lib source                           
#>  assertthat    0.2.1        2019-03-21 [1] CRAN (R 3.6.0)                   
#>  backports     1.1.5        2019-10-02 [1] CRAN (R 3.6.0)                   
#>  bit           1.1-15       2020-01-13 [1] CRAN (R 3.6.1)                   
#>  bit64         0.9-7        2017-05-08 [1] CRAN (R 3.6.0)                   
#>  callr         3.4.0        2019-12-09 [1] CRAN (R 3.6.0)                   
#>  cli           2.0.1        2020-01-08 [1] CRAN (R 3.6.1)                   
#>  crayon        1.3.4        2017-09-16 [1] CRAN (R 3.6.0)                   
#>  desc          1.2.0        2018-05-01 [1] CRAN (R 3.6.0)                   
#>  devtools      2.2.1        2019-09-24 [1] CRAN (R 3.6.0)                   
#>  digest        0.6.23       2019-11-23 [1] CRAN (R 3.6.0)                   
#>  ellipsis      0.3.0        2019-09-20 [1] CRAN (R 3.6.0)                   
#>  evaluate      0.14         2019-05-28 [1] CRAN (R 3.6.0)                   
#>  fansi         0.4.1        2020-01-08 [1] CRAN (R 3.6.1)                   
#>  fs            1.3.1        2019-05-06 [1] CRAN (R 3.6.0)                   
#>  glue          1.3.1        2019-03-12 [1] CRAN (R 3.6.0)                   
#>  highr         0.8          2019-03-20 [1] CRAN (R 3.6.0)                   
#>  htmltools     0.4.0        2019-10-04 [1] CRAN (R 3.6.0)                   
#>  knitr         1.26         2019-11-12 [1] CRAN (R 3.6.0)                   
#>  lifecycle     0.1.0        2019-08-01 [1] CRAN (R 3.6.0)                   
#>  magrittr      1.5          2014-11-22 [1] CRAN (R 3.6.0)                   
#>  memoise       1.1.0        2017-04-21 [1] CRAN (R 3.6.0)                   
#>  pillar        1.4.3.9000   2020-01-07 [1] Github (r-lib/pillar@8f5918c)    
#>  pkgbuild      1.0.6        2019-10-09 [1] CRAN (R 3.6.0)                   
#>  pkgconfig     2.0.3        2019-09-22 [1] CRAN (R 3.6.0)                   
#>  pkgload       1.0.2        2018-10-29 [1] CRAN (R 3.6.0)                   
#>  prettyunits   1.1.0        2020-01-09 [1] CRAN (R 3.6.1)                   
#>  processx      3.4.1        2019-07-18 [1] CRAN (R 3.6.0)                   
#>  ps            1.3.0        2018-12-21 [1] CRAN (R 3.6.0)                   
#>  purrr         0.3.3        2019-10-18 [1] CRAN (R 3.6.0)                   
#>  R6            2.4.1        2019-11-12 [1] CRAN (R 3.6.0)                   
#>  Rcpp          1.0.3        2019-11-08 [1] CRAN (R 3.6.0)                   
#>  remotes       2.1.0        2019-06-24 [1] CRAN (R 3.6.0)                   
#>  rlang         0.4.2.9000   2020-01-14 [1] Github (r-lib/rlang@a669094)     
#>  rmarkdown     2.0          2019-12-12 [1] CRAN (R 3.6.0)                   
#>  rprojroot     1.3-2        2018-01-03 [1] CRAN (R 3.6.0)                   
#>  sessioninfo   1.1.1        2018-11-05 [1] CRAN (R 3.6.0)                   
#>  stringi       1.4.5        2020-01-11 [1] CRAN (R 3.6.1)                   
#>  stringr       1.4.0        2019-02-10 [1] CRAN (R 3.6.0)                   
#>  testthat      2.3.1        2019-12-01 [1] CRAN (R 3.6.0)                   
#>  tibble        2.99.99.9014 2020-01-14 [1] Github (tidyverse/tibble@413e6e4)
#>  tidyselect    0.2.99.9000  2020-01-13 [1] Github (r-lib/tidyselect@b664f08)
#>  usethis       1.5.1        2019-07-04 [1] CRAN (R 3.6.0)                   
#>  vctrs         0.2.99.9002  2020-01-14 [1] Github (r-lib/vctrs@de015f5)     
#>  vroom         1.2.0        2020-01-13 [1] CRAN (R 3.6.1)                   
#>  withr         2.1.2        2018-03-15 [1] CRAN (R 3.6.0)                   
#>  xfun          0.12         2020-01-13 [1] CRAN (R 3.6.1)                   
#>  yaml          2.2.0        2018-07-25 [1] CRAN (R 3.6.0)                   
#> 
#> [1] /Library/Frameworks/R.framework/Versions/3.6/Resources/library
```

</details>

This is probably a problem with ALTREP vectors as the following works

``` r
x <- vroom::vroom("x,y\n1,2\n", altrep = FALSE)
#> Rows: 1
#> Columns: 2
#> Delimiter: ","
#> dbl [2]: x, y
#> 
#> Use `spec()` to retrieve the guessed column specification
#> Pass a specification to the `col_types` argument to quiet this message
vctrs::vec_rbind(x)
#> # A tibble: 1 x 2
#>       x     y
#>   <dbl> <dbl>
#> 1     1     2
```

### PR DESCRIPTION
**Pull request by DavisVaughan**
Closes #744 

Also adds `NA_integer_` subscript support to the `altrep_rle` class so we can add tests, and fixes a small off by one bug with it.

Now, with this PR and dev vroom:

``` r
library(vctrs)
library(vroom)

x <- vroom("x,y\n1,2\n")
#> Rows: 1
#> Columns: 2
#> Delimiter: ","
#> dbl [2]: x, y
#> 
#> Use `spec()` to retrieve the guessed column specification
#> Pass a specification to the `col_types` argument to quiet this message
vec_rbind(x)
#> # A tibble: 1 x 2
#>       x     y
#>   <dbl> <dbl>
#> 1     1     2

x <- vroom("x,y\n1,2\n")
#> Rows: 1
#> Columns: 2
#> Delimiter: ","
#> dbl [2]: x, y
#> 
#> Use `spec()` to retrieve the guessed column specification
#> Pass a specification to the `col_types` argument to quiet this message
vec_init(x, 2)
#> # A tibble: 2 x 2
#>       x     y
#>   <dbl> <dbl>
#> 1    NA    NA
#> 2    NA    NA
```

<sup>Created on 2020-01-14 by the [reprex package](https://reprex.tidyverse.org) (v0.3.0.9000)</sup>

### CODE REVIEW

File: tests/testthat/test-slice.R [hunk 1]
```diff
@@ -806,6 +830,6 @@ test_that("vec_slice() works with Altrep classes with custom extract methods", {
 
   x <- .Call(vctrs_rle, c(foo = 10L, bar = 5L))
 
-  idx <- c(1, 3, 15)
-  expect_equal(vec_slice(x, idx), x[idx])
+  idx <- c(9, 10, 11)
+  expect_equal(vec_slice(x, idx), c("foo", "foo", "bar"))
```
Comment (DavisVaughan) [inline tests/testthat/test-slice.R]: I've tweaked this test to hone in on the location where we switch between foo and bar. Previously it was off by one, and this would catch that now.
